### ISSUE TITLE
Investigate use of MLIL to pick up stack block and byref values

### ISSUE THREAD

**Issue by droe** [droe/binja-blocks#29]
Current approach is based on walking HLIL to find assignments to stack variables, which is brittle for complex functions, especially when the same stack area is used for different purposes in different branches, or when lifting is imperfect (like when involving arm64 floating-point registers).

A better approach might be:
-   Optionally figure out which MLIL instruction(s) represents initialization (for later use in variable feng shui, and as starting point for finding first use)
-   Figure out which MLIL instruction represents first use of a stack block or byref
-   Query MLIL for the values on the stack at the time of first use

If this works, it might increase robustness of the discovery and reading of structures on the stack.  It would not automatically untangle the display in HLIL after annotation with the struct type.

**Comment by WeiN76LQh** [comment]
I agree with this, there are store and load from memory MLIL instructions that are probably what you want. It would certainly improve robustness to try and do most, if not all, using MLIL where possible.

If `__NSConcreteStackBlock` is being assigned to a variable not a struct field (which requires the variable to be typed), like so:
```c
void* (* const var_100)[0x20] = __NSConcreteStackBlock
```
Then MLIL returns the operation `MLIL_SET_VAR`.

Once the plugin has done its magic then you get the following line for the line above:

```c
stack_block_var_100.isa = __NSConcreteStackBlock
```
MLIL returns the operation `MLIL_SET_VAR_FIELD`.

So you'd probably just need to look out for those 2 operations. Not sure if `MLIL_STORE` or `MLIL_STORE_STRUCT` could occur as well.

Also I think this might be related to the use of HLIL by the plugin but I'm not sure. There is alot of calls to `update_analysis_and_wait` by the plugin which slows it down alot. It would complete alot faster without them but I'm not sure why they're necessary? I kind of assumed it was something to do with getting the right information from HLIL.

**Comment by droe** [comment]
I was hoping to get some assistance from Binary Ninja's tracking of constants written to the stack.

What does not seem to work (as in, results are always `<undetermined>`):
-   `get_stack_contents`
-   `get_possible_stack_contents`
-   `get_ssa_var_possible_values`

Tested with both `arm64` and `x86_64`, even for the simplest of cases, these approaches don't give any results other than undetermined values.  Not sure if I'm holding it wrong, my expectations are wrong, or something is broken with Binja, possibly only for MachO/Apple targets.

**Comment by WeiN76LQh** [comment]
I think without it its getting to get more difficult. I'm experiencing the same issue, although my testing wasn't very thorough. Hopefully you get an answer on Slack
